message_edit: Don't display resolved topics as MOVED.

timabbott opened the pull request:
The correct algorithm here is to loop through edit history entries,
and see if any of them are more interesting than a (un)resolve topic
edit, extending the existing loop we had.

Untested because this doesn't work -- I'd like us to fix the server API
first.

Fixes #19919.

laurynmm commented:
@timabbott - Here's where I've gotten so far...

There is still a live update bug due to the fact that the `update_messages` event code simulates the server generated edit_history logs, but in doing so creates two separate logs for an edit that includes both a topic and content edit. The topic edit log happens first and the page is narrowed. But a similar page update doesn't happen for the content edit.

Also, I found another bug where a non-existent event field was being used for the edit log `user_id`. I made a comment, but haven't thought through the best fix yet.

Finally, the node test could probably be improved, but I at least updated it to be testing the changes so far.

timabbott commented on frontend_tests/node_tests/message_list_view.js:
```diff
@@ -99,47 +99,58 @@ test("msg_moved_var", () => {
             }),
-            // topic, stream, and content changed: Edit
+            // topic unresolved and content changed: EDITED
+            // added content change to get the container field
```
I guess we need to add a test case for "only resolved", right?

laurynmm replied:
Yes, I expanded the tests in the updated commit so that all the single edit possibilities are tested as well as two multiple edit history entries. There is one test - for a topic unresolved - that should never happen because it doesn't also have an associated topic resolved log, but I thought we should just test the logic of the loop just in case ...

timabbott replied:
Yep, I think this is the right testing strategy.

timabbott commented on static/js/message_events.js:
```diff
@@ -234,11 +234,18 @@ export function update_messages(events) {
                      * without a browser reload. */
                     const edit_history_entry = {
+                        // BUG: event.edited_by field does not exist
                         edited_by: event.edited_by,
```
I think `event.user_id` does exist and we can use that:
```
    event: Dict[str, Any] = {                                                                            
        "type": "update_message",                                                                        
        "user_id": user_profile.id,                                                                      
        "edit_timestamp": datetime_to_timestamp(timestamp),                                              
        "message_id": target_message.id,                                                                 
        "rendering_only": False,                                                                         
    }                                                                                                    
                                                                                                         
    edit_history_event: EditHistoryEvent = {                                                             
        "user_id": user_profile.id,                                                                      
        "timestamp": event["edit_timestamp"],                                                            
    }                                                                                                    

```

timabbott commented on static/js/message_events.js:
```diff
@@ -395,6 +402,7 @@ export function update_messages(events) {
                 // two client-side events, which is probably good for display.
                 const edit_history_entry = {
+                    // BUG: event.edited_by field does not exist
                     edited_by: event.edited_by,
```
Maybe let's fix the `edited_by` bug in a quick prep commit? It seems orthogonal and also kinda embarrassing :)

timabbott commented on static/js/message_list_view.js:
```diff
@@ -57,22 +57,53 @@ function same_recipient(a, b) {
+                // it was a true move or a resolve/unresolve.
+                if (
+                    edit_history_event.topic.startsWith(resolved_topic.RESOLVED_TOPIC_PREFIX) &&
+                    edit_history_event.topic.slice(2) === edit_history_event.prev_topic
```
I think these can use the new in the last week `static/shared/js/resolved_topics` library for these checks, and we should probably do that to avoid an abstraction violation.

laurynmm replied:
Yes, I should have looked more carefully at that when I did the rebase yesterday.

timabbott commented on frontend_tests/node_tests/message_list_view.js:
```diff
@@ -99,47 +99,58 @@ test("msg_moved_var", () => {
+                ],
             }),
-            // content edited: Edit
+            // content edited: EDITIED
```
Small typo here

laurynmm commented:
@timabbott  - The updated commit addresses all of your review notes, as well as the content + topic edit bug.

To fix that I basically moved the chunk of code where we simulate the edit history entry for the content change to happen before the topic / stream change loop. That way the content edit exists before the topic edit, and when the topic edit loop does the narrowing for the live view, both edit history entries exist, with the content edit getting priority for the 'EDITED' label.

And since the topic loop sets the `last_edit_timestamp` for the message, that happens before the potential narrowing as well. And if it's a content only edit, the topic loop is skipped and we immediately check the `rendering_only` boolean for setting the `last_edit_timestamp`.

I don't know whether there's a need to do some assertion at some point that if there is a `last_edit_timestamp` on the stored  message, then there is also an `edit_history` as well. It occurred to me, but it also doesn't seem really necessary. I did make it so that the test followed that assertion because it bothered me it didn't before.

Other than the updates to the tests I've done a pretty through manual testing and it all looks good.

timabbott commented:
This is awesome, merged with these small changes. The first is swapping a conditional order to feel perhaps a bit more intuitive, and the second just extending a comment for readability.  Huge thanks for doing this project @laurynmm!

```
diff --git a/static/js/message_list_view.js b/static/js/message_list_view.js
index 9870a09903..06bda51f4e 100644
--- a/static/js/message_list_view.js
+++ b/static/js/message_list_view.js
@@ -283,13 +283,14 @@ export class MessageListView {
         const edit_history_details = analyze_edit_history(message_container.msg);
 
         if (
-            !(edit_history_details.moved || edit_history_details.edited) ||
-            last_edit_timestr === undefined
+            last_edit_timestr === undefined ||
+            !(edit_history_details.moved || edit_history_details.edited)
         ) {
             // For messages whose edit history at most includes
-            // resolving topics, we don't display anything here (The
-            // message actions popover will still display an edit
-            // history option).
+            // resolving topics, we don't display an EDITED/MOVED
+            // notice at all. (The message actions popover will still
+            // display an edit history option, so you can see when it
+            // was marked as resolved if you need to).
             delete message_container.last_edit_timestr;
             message_container.edited_in_left_col = false;
             message_container.edited_alongside_sender = false;
```
